File: CBORSingle.cs
/*
 * Created by SharpDevelop.
 * User: Peter
 * Date: 2/21/2014
 * Time: 9:26 PM
 * 
 * To change this template use Tools | Options | Coding | Edit Standard Headers.
 */
using System;

namespace PeterO
{
  internal class CBORSingle : ICBORNumber
  {
    
    public bool IsPositiveInfinity(object obj)
    {
      throw new NotImplementedException();  // TODO: Implement
    }
    
    public bool IsInfinity(object obj)
    {
      throw new NotImplementedException();  // TODO: Implement
    }
    
    public bool IsNegativeInfinity(object obj)
    {
      throw new NotImplementedException();  // TODO: Implement
    }
    
    public bool IsNaN(object obj)
    {
      throw new NotImplementedException();  // TODO: Implement
    }
    
    public double AsDouble(object obj)
    {
      return (double)(float)obj;
    }
    
    public ExtendedDecimal AsExtendedDecimal(object obj)
    {
      throw new NotImplementedException();  // TODO: Implement
    }
    
    public ExtendedFloat AsExtendedFloat(object obj)
    {
      throw new NotImplementedException();  // TODO: Implement
    }
    
    public float AsSingle(object obj)
    {
      throw new NotImplementedException();  // TODO: Implement
    }
    
    public BigInteger AsBigInteger(object obj)
    {
      throw new NotImplementedException();  // TODO: Implement
    }
    
    public long AsInt64(object obj)
    {
      float fltItem = (float)obj;
      if (Single.IsNaN(fltItem)) {
        throw new OverflowException("This object's value is out of range");
      }
      fltItem = (fltItem < 0) ? (float)Math.Ceiling(fltItem) : (float)Math.Floor(fltItem);
      if (fltItem >= Int64.MinValue && fltItem <= Int64.MaxValue) {
        return (long)fltItem;
      }
      throw new OverflowException("This object's value is out of range");
    }
    
    public bool CanFitInSingle(object obj)
    {
      return true;
    }
    
    public bool CanFitInDouble(object obj)
    {
      return true;
    }
    
    public bool CanFitInInt32(object obj)
    {
      throw new NotImplementedException();  // TODO: Implement
    }
    
    public bool CanFitInInt64(object obj)
    {
      throw new NotImplementedException();  // TODO: Implement
    }
    
    public bool CanTruncatedIntFitInInt64(object obj)
    {
      throw new NotImplementedException();  // TODO: Implement
    }
    
    public bool CanTruncatedIntFitInInt32(object obj)
    {
      throw new NotImplementedException();  // TODO: Implement
    }
    
    public int AsInt32(object obj, int minValue, int maxValue)
    {
      throw new NotImplementedException();  // TODO: Implement
    }
    
    public bool IsZero(object obj)
    {
      throw new NotImplementedException();  // TODO: Implement
    }
    
    public int Sign(object obj)
    {
      throw new NotImplementedException();  // TODO: Implement
    }
    
    public bool IsIntegral(object obj)
    {
      throw new NotImplementedException();  // TODO: Implement
    }
    
  public bool CanFitInTypeZeroOrOne(object obj)
  {
    throw new NotImplementedException();
  }
  }
}

File: src/com/upokecenter/util/CBORSingle.java
package com.upokecenter.util;
/*
Written in 2014 by Peter O.
Any copyright is dedicated to the Public Domain.
http://creativecommons.org/publicdomain/zero/1.0/
If you like this, you should donate to Peter O.
at: http://peteroupc.github.io/CBOR/
 */

  final class CBORSingle implements ICBORNumber
  {
    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A Boolean object.
     */
    public boolean IsPositiveInfinity(Object obj) {
      return ((((Float)obj).floatValue())==Float.POSITIVE_INFINITY);
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A Boolean object.
     */
    public boolean IsInfinity(Object obj) {
            return ((Float)(((Float)obj).floatValue())).isInfinite();
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A Boolean object.
     */
    public boolean IsNegativeInfinity(Object obj) {
      return ((((Float)obj).floatValue())==Float.NEGATIVE_INFINITY);
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A Boolean object.
     */
    public boolean IsNaN(Object obj) {
      return Float.isNaN(((Float)obj).floatValue());
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A 64-bit floating-point number.
     */
    public double AsDouble(Object obj) {
      return ((Float)obj).doubleValue();
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return An ExtendedDecimal object.
     */
    public ExtendedDecimal AsExtendedDecimal(Object obj) {
      return ExtendedDecimal.FromSingle(((Float)obj).floatValue());
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return An ExtendedFloat object.
     */
    public ExtendedFloat AsExtendedFloat(Object obj) {
      return ExtendedFloat.FromSingle(((Float)obj).floatValue());
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A 32-bit floating-point number.
     */
    public float AsSingle(Object obj) {
      return ((Float)obj).floatValue();
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A BigInteger object.
     */
    public BigInteger AsBigInteger(Object obj) {
      return CBORUtilities.BigIntegerFromSingle(((Float)obj).floatValue());
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A 64-bit signed integer.
     */
    public long AsInt64(Object obj) {
      float fltItem = ((Float)obj).floatValue();
      if (Float.isNaN(fltItem)) {
        throw new ArithmeticException("This Object's value is out of range");
      }
      fltItem = (fltItem < 0) ? (float)Math.ceil(fltItem) : (float)Math.floor(fltItem);
      if (fltItem >= Long.MIN_VALUE && fltItem <= Long.MAX_VALUE) {
        return (long)fltItem;
      }
      throw new ArithmeticException("This Object's value is out of range");
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A Boolean object.
     */
    public boolean CanFitInSingle(Object obj) {
      return true;
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A Boolean object.
     */
    public boolean CanFitInDouble(Object obj) {
      return true;
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A Boolean object.
     */
    public boolean CanFitInInt32(Object obj) {
      throw new UnsupportedOperationException();  // TODO: Implement
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A Boolean object.
     */
    public boolean CanFitInInt64(Object obj) {
      throw new UnsupportedOperationException();  // TODO: Implement
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A Boolean object.
     */
    public boolean CanTruncatedIntFitInInt64(Object obj) {
      throw new UnsupportedOperationException();  // TODO: Implement
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A Boolean object.
     */
    public boolean CanTruncatedIntFitInInt32(Object obj) {
      throw new UnsupportedOperationException();  // TODO: Implement
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @param minValue A 32-bit signed integer. (2).
     * @param maxValue A 32-bit signed integer. (3).
     * @return A 32-bit signed integer.
     */
    public int AsInt32(Object obj, int minValue, int maxValue) {
      throw new UnsupportedOperationException();  // TODO: Implement
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A Boolean object.
     */
    public boolean IsZero(Object obj) {
      throw new UnsupportedOperationException();  // TODO: Implement
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A 32-bit signed integer.
     */
    public int Sign(Object obj) {
      throw new UnsupportedOperationException();  // TODO: Implement
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A Boolean object.
     */
    public boolean IsIntegral(Object obj) {
      throw new UnsupportedOperationException();  // TODO: Implement
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object.
     * @return A Boolean object.
     */
    public boolean CanFitInTypeZeroOrOne(Object obj) {
      throw new UnsupportedOperationException();
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object. (2).
     * @return An arbitrary object.
     */
public Object Negate(Object obj) {
      float val = ((Float)obj).floatValue();
      return -val;
    }

    /**
     * Not documented yet.
     * @param obj An arbitrary object. (2).
     * @return An arbitrary object.
     */
public Object Abs(Object obj) {
      float val = ((Float)obj).floatValue();
      return (val < 0) ? -val : obj;
    }
  }
